File: src/main/java/custom/tibame201020/adbOpenCv/script/gearScript/gear/GearUpgradeAdapter.java
package custom.tibame201020.adbOpenCv.script.gearScript.gear;

import java.util.List;

public interface GearUpgradeAdapter {

    boolean isOkToUpgrade(
            GearDTOs.GearSet gearSet,
            GearDTOs.GearType gearType,
            int currentLevel,
            GearDTOs.GearProp gearProp
    );

    boolean isOkToStore(
            GearDTOs.GearSet gearSet,
            GearDTOs.GearType gearType,
            GearDTOs.GearProp gearProp
    );

    List<GearDTOs.GearSet> GEAR_SET_FOUR_TYPES = List.of(
            GearDTOs.GearSet.ATTACK,
            GearDTOs.GearSet.DESTRUCTION,
            GearDTOs.GearSet.SPEED,
            GearDTOs.GearSet.RAGE,
            GearDTOs.GearSet.REVENGE,
            GearDTOs.GearSet.COUNTER,
            GearDTOs.GearSet.INJURY,
            GearDTOs.GearSet.LIFE_STEAL,
            GearDTOs.GearSet.PROTECTION
    );

    List<GearDTOs.GearSet> GEAR_SET_TWO_TYPES = List.of(
            GearDTOs.GearSet.DEFENSE,
            GearDTOs.GearSet.HEALTH,
            GearDTOs.GearSet.HIT,
            GearDTOs.GearSet.RESISTANCE,
            GearDTOs.GearSet.CRITICAL,
            GearDTOs.GearSet.DUAL_ATTACK,
            GearDTOs.GearSet.IMMUNITY,
            GearDTOs.GearSet.PENETRATION,
            GearDTOs.GearSet.TORRENT

    );

    default GearDTOs.GearSetType gearSetType(GearDTOs.GearSet gearSet) {
        if (GEAR_SET_FOUR_TYPES.contains(gearSet)) {
            return GearDTOs.GearSetType.FOUR;
        }
        if (GEAR_SET_TWO_TYPES.contains(gearSet)) {
            return GearDTOs.GearSetType.TWO;
        }
        throw new RuntimeException("un caught gear set type");
    }
}

File: py-version/script/gearScript/gear/gear_upgrade_adapter.py
"""
Gear 升級適配器介面
"""
from abc import ABC, abstractmethod
from typing import List
from .gear_dtos import Gear, GearSet, GearSetType


class GearUpgradeAdapter(ABC):
    """裝備升級適配器抽象類別"""
    
    GEAR_SET_FOUR_TYPES = [
        GearSet.ATTACK, GearSet.DESTRUCTION, GearSet.SPEED, GearSet.RAGE,
        GearSet.REVENGE, GearSet.COUNTER, GearSet.INJURY, GearSet.LIFE_STEAL,
        GearSet.PROTECTION
    ]
    
    GEAR_SET_TWO_TYPES = [
        GearSet.DEFENSE, GearSet.HEALTH, GearSet.HIT, GearSet.RESISTANCE,
        GearSet.CRITICAL, GearSet.DUAL_ATTACK, GearSet.IMMUNITY,
        GearSet.PENETRATION, GearSet.TORRENT
    ]
    
    @abstractmethod
    def is_ok_to_upgrade(self, gear: Gear) -> bool:
        """判斷是否可以升級"""
        pass
    
    @abstractmethod
    def is_ok_to_store(self, gear: Gear) -> bool:
        """判斷是否可以儲存"""
        pass
    
    def gear_set_type(self, gear_set: GearSet) -> GearSetType:
        """取得套裝類型"""
        if gear_set in self.GEAR_SET_FOUR_TYPES:
            return GearSetType.FOUR
        if gear_set in self.GEAR_SET_TWO_TYPES:
            return GearSetType.TWO
        raise RuntimeError("un caught gear set type")


